### ISSUE TITLE
[Camel] Camel integration in the Kura CloudService

### ISSUE THREAD

**Issue by cdealti** [eclipse/kura#189]
@hekonsek this feature leverages the Kura multiple connections feature recently added on the develop branch.

You need to implement a new [CloudServiceFactory](https://github.com/eclipse/kura/blob/develop/kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/cloud/factory/CloudServiceFactory.java) and a component registering a `CloudService` in the OSGi framework.

The Camel implementations will be `CamelCloudServiceFactory` and `CamelCloudService`.

The Kura [default](https://github.com/eclipse/kura/blob/develop/kura/org.eclipse.kura.core.cloud/src/main/java/org/eclipse/kura/core/cloud/factory/DefaultCloudServiceFactory.java) implementation of the `CloudServiceFactory` can be used as a reference implementation.

Please note that the CloudService implementation must also be a Kura `ConfigurableComponent` so, in the case of the `CamelCloudService`, its configuration could be the Camel XML routes for that CloudService instance (the same configuration of the Camel quickstart example).

**Comment by ctron** [comment]
@cdealti Maybe you can help a little bit here. I am trying to look for code referencing the `CloudServiceFactory` interface. And to me it seems as if it is not really used in Kura. There is the "default" implementation. But not more. No one seems to call or use the interface. Unless I have missed a spot.

**Comment by cdealti** [comment]
@ctron it is mostly intended for the Web UI so that it can get registered ``CloudServiceFactory`` OSGi services and present a consistent UI to create connection instances and show their configurations.

Can you create an issue for this? See also #444

**Comment by ctron** [comment]
I have to admit I fail to see the basic idea behind the way this is done. On the one hand there is a common interface, on the other hand it is not used but every component has to create it own way of adding services.

Maybe you have some description, beside the Javadoc, what the intended use cases of this construct is. As far as I understood, once the method is called in order to create that service, it still has to be configured. So instead of making a custom console command, as described in issue #444, shouldn't there be a common console command providing this functionality? Or did I misunderstand that?

**Comment by cdealti** [comment]
> I have to admit I fail to see the basic idea behind the way this is done. On the one hand there is a common interface, on the other hand it is not used but every component has to create it own way of adding services.

What do you exactly mean?

> Maybe you have some description, beside the Javadoc, what the intended use cases of this construct is.

The Javadoc and the [Multiple Cloud Connections Wiki page](https://github.com/eclipse/kura/wiki/Multiple-Cloud-Connections) are the only source of information about this topic.

The intended use is providing a service interface that the Web UI would use to allow to:
1. Discover registered ``CloudServiceFactory`` services, for example the Kura ``DefaultCloudServiceFactory`` or the ``WatsonIoTCloudServiceFactory``.
2. Present an unified UI to create or remove new connections from the discovered Cloud Service Factories. For example selecting a registered Cloud Service Factory and clicking on a *New* button.

> As far as I understood, once the method is called in order to create that service, it still has to be configured. So instead of making a custom console command, as described in issue #444, shouldn't there be a common console command providing this functionality? Or did I misunderstand that?

You are right. However the gogo shell command was only added for testing purposes because the Web UI does not yet allow the flow described above. On the other hand, adding the gogo shell commands only requires adding a couple of properties to the component definition and no code.

Your are also right on the need of configuring the connection after its creation.
Currently, when you create a new connection in Kura, three new services will appear in the Web UI Services panel.
The idea was to remove connection-related services from here and moving them to a dedicated part of the UI.
Note that some connection implementations are layered (for example the Kura one) and the overall configuration is made by three separate configurations.
We wanted to group the configurations of the three layers together which can be done using existing APIs of the ``ConfigurationService`` or by some convention still to be decided.

I believe that we should spend more thoughts on these topics when we will have the Watson IoT connections implementation (which is also a three layer implementation, reusing the Kura DataService and MqttTransportService implementations and implementing its own CloudService).

**Comment by ctron** [comment]
What I mean is, that currently there is an interface `CloudServiceFactory`, which is never used in the Web UI. The create command `kura.cloud:createConfiguration`, which suggests being generic (by its naming) however is directly implemented by `DefaultCloudServiceFactory`, without using any interface. So the only possible way, as of now, is to make a complete manual implementation. Not using the interface and not using some sort of generic console command. And even after the creation, everything still has to be configured in a way that is not covered by the interfaces, but goes with the `ConfigurableComponent` construct.

Wouldn't it make more sense to actually implement some sort of management service, hosting the configurations for different implementations and then leveraging the `CloudServiceFactory` to create instances based on those configurations. That way it would be possible to track the existing configurations (there would be some sort of status, even if the component did not register), there could be a state for missing components, …

Or at least provide a console command or Web UI to support this new construct.

**Comment by ctron** [comment]
This one is covered now the PR #472

**Comment by cdealti** [comment]
I've fetched #472 and started it in the emulator.
I think I need a few clarifications:

1. It looks like ``CloudServiceFactory`` is not used and you implemented a parallel method to create and delete ``CamelCloudService``S? I'm not saying this is wrong and I understand you don't agree on the design of the Cloud Service Factory but I don't think this is a good reason to "boycott" it.
We are open to backward compatible changes to the Cloud Service Factory API if you think this is needed but I don't understand why it cannot be used for the Camel Cloud Service. Please consider that:
    * It's already implemented by Kura
    * IBM will contribute an WatsonIoTCloudService implementation based on the Cloud Service Factory. See #444. As far as I know this will be showcased by Patrizia Gufler at [IFA 2016](https://www.youtube.com/watch?v=wZoF58Hn18w)
    * We will need to implement two different methods to discover, create and delete Cloud ServiceS in the Web UI.

2. In your design I've noticed that you can create a CamelCloudService configuration specifying a service ranking. I understand this is related to #466. As agreed I've commented the issue and updated the Wiki page with a simple proposal that should fulfill your concerns.

3. After creating a Camel Cloud Service configuration I cannot find a new ``CloudService`` registered in the framework with the specified pid. I have to take a closer look tomorrow.

**Comment by ctron** [comment]
1. I don't try to boycott it, but I also try to get some functionality out of that implementation. Right now if I do implement that interface I cannot make use of that feature. So I decided to add an additional layer to actively provide an interface which does enable a user to make use of that functionality. I can make an implementation of that interface rather quickly, but then again, it simply is not being used by Kura. So for the beginning I did focus on a way which allows me to actually use that feature. But I can make an additional implementation of that feature if that helps you.

2. I do like the proposal that you made. It looks really good and the service ranking also does not work in all cases. But helped my testing the feature for the moment. I am happy to remove it once the Kura example services are migrated to your new concept.

3. That is true. And it is not perfect. However having the user enter a PID can currently crash the Web UI. In addition there are a bunch of other assumptions that the UI makes in order to create a label for the component. So instead of tracing down and changing all that behavior in the UI, I let the  component adapt to the Web UI in order to prevent the issues that I encountered. The final service.pid should be the prefix `org.eclipse.kura.camel.cloud.factory.CamelFactory-` and the suffix `-CloudService`. I thought about letting the user enter additional properties which simply get registered with OSGi, like the service ranking, but I am not sure this is a good idea. But I could register the entered, short, PID in addition under a different property.

So I think the component does what it should do, provide an implementation of CloudService backed by Camel with XML. I will make an additional implementation of the `CloudServiceFactory` since that seems to be biggest issue from your side.

**Comment by ctron** [comment]
So the interface is implemented.

**Comment by cdealti** [comment]
> I don't try to boycott

Clearly mine was a joke and "boycott" was between quotes. I didn't intend to flame the discussion.

Let's try to fix 3 and everything is needed to make the experience consistent.

I just want to stress that I agree that the current design is not perfect but dictated by the need to quickly respond to specific customer use cases in a backward compatible way.

I also like your idea of having a centralized service for the Cloud Service Factory. I believe at some point this will be implemented in a specific part of the Web UI.

**Comment by ctron** [comment]
:wink: 

So I could also got that way with point 3) to simply fail first, if the user tries to create something which is not possible, then I fail when the component is created. That would still let the user run into an error, but would be consistent from a PID perspective. And once the Web UI is fixed, we can remove that.

**Comment by cdealti** [comment]
About item 3 above and the current UI limitations... When we added the Cloud Service Factory we had in mind to create some other things. Informally these are:

a. The Configurable Components created by Cloud Service Factory will not be shown as plain **Services** in the Web UI. They will be shown in a dedicated **Connections** part of the Web UI.

a.1 This **Connections** section will get all the registered Cloud Connection Factories and provide a drop down list with their pids. A "New" button allows to create a new configuration (a new connection) for the selected factory. A "Delete" button will delete an existing configuration (an existing configuration).

b. (Nasty). Have you noticed that when you create a new Kura connection from the ``DefaultCloudServiceFactory`` three new services are created? This happens because the classic Kura connection stack had originally three layers, a Cloud Service, a Data Service and an Data Transport Service, each being an Configurable Component. The need to create three configurations and link them together is the main rationale for the existence of the Cloud Service Factory. This is clearly described in the Javadoc but I agree it's not the neatest design. We decided to make the minimal changes to allow multiple stack instance without code changes and relying on OSGi capabilities (the infamous ``target`` filter/property). Pragmatically I believe this has been a win.

c. If we need to move these connection-related services as described in a) we have now a few requirements/issues:

c.1. We want a way to show the configuration of the 3-layer Kura default stack(s) in a "prettier" way than the generic one displayed for a generic Configurable Component whose configuration is shown in the **Services** section of the Web UI. The **Services** section should be used as the default place allowing to change the configuration if a more specialized "widget" cannot be found.

c.2 We want to support showing configurations for other Cloud Service implementations, like the Camel and the WatsonIoT one. If we cannot find a generic way to do that we must allow a Cloud Service provider to plug the widget rendering its configuration. I think I have some ideas on a generic method to avoid the burden of the pluggable approach.

The above plan has not been implemented yet. I wonder if you agree on that.

**Comment by ctron** [comment]
I would take a different approach. I do understand that the existing interfaces and services should stay present, as they are required by existing application logic. However you are mixing Web UI related issues with the backend architecture.

I don't think it is a limitation to have three services backup up your backend stack. But I agree that, from a UI perspective this is not something which is easily understandable for an end user.

So instead of adding a bunch of implicit dependencies between those components, I would add a "virtual" service (as ConfigurableComponent) which holds the information regarding the service you want to set up (no matter if this will be one or five services in the end). That service is what you configure in the UI. Once this service is started, with will itself create, start and register all required services in the background, taking care of those 3 services in your case.

The benefit in that is, that you have one single component managing those three components. You can check in advance if the configuration is valid and maybe, in the future, even provide some sort of status, validation information about that connection in a single place. Right now if you don't enter a valid MQTT client ID, you just see an exception in the log.

**Comment by ctron** [comment]
I just updated the PR in order to pass through the user entered PID to the CloudService.

**Comment by cdealti** [comment]
The following mockup illustrates how we planned to modify the Web UI to support multiple CloudService providers (Kura default, Camel, WatsonIoT etc.). It's still based on the classic web UI but it gives the feeling:

![image](https://cloud.githubusercontent.com/assets/1225835/18269129/1cdab368-7427-11e6-856c-b1b7aba5ceaf.png)

Your PR implements things in a different way. There is a new Camel Cloud Factory in the Service panel to create a new Camel Cloud Service:

![image](https://cloud.githubusercontent.com/assets/1225835/18269435/81bc99b2-7428-11e6-8fad-731e0586f0fd.png)

When the configuration is applied a new service appears that can be updated or deleted

![image](https://cloud.githubusercontent.com/assets/1225835/18269533/cf4eafda-7428-11e6-8721-f59f5b513a07.png)

It's clever but the idea to move cloud related services in a single, consistent user interface.

**Comment by ctron** [comment]
I completely understand the idea behind that. But currently the Web UI does not support that. So we can remove that behavior once the new Web UI is available. Do you have a timeframe for that?

**Comment by cdealti** [comment]
@ctron Let's keep this open as a reminder to remove the custom creation service when we will have #515.

**Comment by ctron** [comment]
I would rather like to create a new one with an explicit title of this task.

**Comment by cdealti** [comment]
@ctron this will be closed once #704 is merged and the Camel Cloud Factory is removed from the generic Web UI Services

**Comment by ctron** [comment]
Exactly, so after all the pending PRs around the Web UI are merged into the release branch, I will issue another PR which removes the custom Web UI.

**Comment by cdealti** [comment]
@ctron Will be closed by #810?

**Comment by ctron** [comment]
Yes. And #810 is merged now.
